symengine/parser.cpp

```diff
@@ -0,0 +1,327 @@
+#include <symengine/symbol.h>
+#include <symengine/add.h>
+#include <symengine/integer.h>
+#include <symengine/rational.h>
+#include <symengine/complex.h>
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/functions.h>
+#include <symengine/constants.h>
+#include <symengine/visitor.h>
+#include <vector>
+#include <stack>
+#include <map>
+#include <set>
+#include <string>
+#include <functional>
+
+namespace SymEngine {
+
+class ExpressionParser {
+
+    // OPERATORS and op_precedence used internally, for parsing
+    std::set<char> OPERATORS = {'-', '+', '*', '/', '^', '(', ')', ','};
+    std::map<char, int> op_precedence = {
+        {')', 0}, {',', 0}, {'-', 1},  {'+', 1},
+        {'*', 3}, {'/', 4}, {'^', 5}
+    };
+    // symengine supported constants
+    std::map<std::string, RCP<const Basic> > constants = {
+
+        {"e", E}, {"E", E}, {"EulerGamma", EulerGamma}, {"pi", pi}, {"I", I}
+    };
+
+    // reference : http://stackoverflow.com/questions/30393285/stdfunction-fails-to-distinguish-overloaded-functions
+    typedef RCP<const Basic> (*single_arg_func)(const RCP<const Basic>&);
+    typedef RCP<const Basic> (*double_arg_func)(const RCP<const Basic>&, const RCP<const Basic>&);
+
+    // cast overlaoded functions below to single_arg, double_arg before they can be used in the map
+    single_arg_func single_casted_log = log;
+    single_arg_func single_casted_zeta = zeta;
+
+    double_arg_func double_casted_log = log;
+    double_arg_func double_casted_zeta = zeta;
+
+    // maps string to corresponding single argument function
+    std::map<   std::string,
+                std::function<RCP<const Basic>(const RCP<const Basic>&)>
+            >   single_arg_functions = {
+
+        {"", [](const RCP<const Basic>& x){return x;}},
+
+        {"sin", sin}, {"cos", cos}, {"tan", tan},
+        {"cot", cot}, {"csc", csc}, {"sec", sec},
+
+        {"asin", asin}, {"acos", acos}, {"atan", atan},
+        {"asec", asec}, {"acsc", acsc}, {"acot", acot},
+
+        {"sinh", sinh}, {"cosh", cosh}, {"tanh", tanh},
+        {"coth", coth}, // {"sech", sech}, {"csch", csch},
+
+        {"asinh", asinh}, {"acosh", acosh}, {"atanh", atanh},
+        {"asech", asech}, {"acoth", acoth}, // {"acsch", acsch},
+
+        {"gamma", gamma}, {"sqrt", sqrt}, {"abs", abs}, {"exp", exp},
+        {"lambertw", lambertw}, {"dirichlet_eta", dirichlet_eta},
+        {"ln", single_casted_log}, {"log", single_casted_log}, {"zeta", single_casted_zeta}
+    };
+
+    // maps string to corresponding double argument function
+    std::map<   std::string,
+                std::function<RCP<const Basic>(const RCP<const Basic>&, const RCP<const Basic>&)>
+            >   double_arg_functions = {
+
+        {"pow", pow}, {"beta", beta}, {"log", double_casted_log}, {"zeta", double_casted_zeta},
+        {"lowergamma", lowergamma}, {"uppergamma", uppergamma}, {"polygamma", polygamma},
+        {"kronecker_delta", kronecker_delta}
+
+    };
```

Comment:
All these can probably be made `const`, since they never change once they are declared.